stackslib/src/chainstate/nakamoto/mod.rs

```diff
@@ -1381,7 +1416,12 @@ impl NakamotoChainState {
         // though it will always be None), which gets the borrow-checker to believe that it's safe
         // to access `stacks_chain_state` again.  In the `Ok(..)` case, it's instead sufficient so
         // simply commit the block before beginning the second transaction to mark it processed.
-        let (ok_opt, err_opt) = match NakamotoChainState::append_block(
+
+        // set the sortition tx's tip to the burnchain view -- we must unset this after appending the block,
+        //  so we wrap this call in a closure to make sure that the unsetting is infallible
+        let prior_sort_tip =
```

Comment:
This is pretty dangerous.  Can you help me understand what's going on here, and why you need to modify the sortition handle transaction's inner context?  That handle is opened to the Nakamoto chains coordinator's current canonical sortition tip.  If there are no bugs, then `burnchain_view_sortid` is going to be an ancestor sortition ID of this tip.

Instead of directly manipulating its state here, can you add a method to `SortitionHandleTx` that takes a target ancestor sortition ID and a closure, and evaluates the closure with `self.context.chain_tip` set to the ancestor sortition ID (much like you're doing here, but in an instance method)?  In addition, please make this method check that the given tip is actually an ancestor of `self.context.chain_tip`.  You can call it something like 

```rust
pub fn at_ancestor<F, R>(&mut self, ancestor_sort_id: &SortitionId, func: F) -> Result<R, db_error>
where
   F: FnOnce(&mut SortitionHandleTx) -> R
```

Reply:
I think if we wanted to avoid modifying the context, I'd prefer to just use a `HandleConn` here instead of a `Tx`. Maybe I can just open a new `Conn` with the `Tx`, or we can avoid the `Tx` and just pass in a `Conn` to begin with (because the only sortition db writes should come after the block processing anyways).

The reason this is necessary is that it forces the clarity instance's burnchain view to be tipped at the burn view. Without this, it could query the canonical tip, which isn't the correct thing to do during block processing.

Reply:
Yup, totally understand why it's necessary.  I'm just trying to find a safer (more maintainable) way to do this :+1: